rfc/0002-report-mode.md

```diff
@@ -0,0 +1,122 @@
+   the regular functionality of the cluster.
+
+## Examples / User Stories
+[examples]: #examples
+
+> Examples of how the feature will be used. Interactions should show the action and the response.
+> When appropriate, provide user stories in the form of "As a [role], I want [feature], so [that]."
+
+As a Kubernetes administrator I want to deploy new policies in a state that they don't perform any
+request rejections, but only inform about what would have been rejected had the policy been active.
+
+As a Kubernetes administrator I want to deploy new policies in "reporting" mode, and then flip the
+switch of the policy to activate it after it has been in "reporting" mode for a short while that I
+considered enough to learn if it would have an undesired impact in the cluster.
+
+As a Kubernetes administrator I want to plot in Grafana acceptions/rejections/mutations coming from
+policies in "report" mode in the same way that I do with policies in "enforce" mode -- the only mode
+Kubewarden had until now.
+
+# Detailed design
+[design]: #detailed-design
+
+> This is the bulk of the RFC. Explain the design in enough detail for somebody familiar with the
+> product to understand, and for somebody familiar with the internals to implement.
+
+> This section should cover architecture aspects and the rationale behind disruptive technical
+> decisions (when applicable), as well as corner-cases and warnings.
+
+## "report" mode
+
+### CRD
+
+The `ClusterAdmissionPolicy` CRD will be extended so it includes a `mode` attribute:
+
+```go
+type ClusterAdmissionPolicySpec struct {
+   // ...
+   Mode string `json:"mode,omitempty"`
+   // ...
+}
+```
+
+Two values are allowed: `enforce` and `report`. If `mode` is omitted, it will be defaulted to
+`enforce`. It's possible to update the `mode` from `report` to `enforce`. Allowing to change `mode`
+from `enforce` to `report` requires further consideration -- so it's not possible to deactivate
+policies just by having `UPDATE` permissions.
+
+### Policy Server
+
+The Policy Server has to be updated so the `mode` is included in the [`Policy` struct used for
+settings](https://github.com/kubewarden/policy-server/blob/c8d64da87448b7f9250a1d6b5e56817f25b56359/src/settings.rs#L11-L19).
+
+The Policy Server will always accept the request if this struct `mode` is set to `report`.
+
+The Policy Server will log with INFO level every evaluation specifying the `mode` of the policy. If
+the policy is mutating the resource, the resulting object will be printed in the structured logging
+output.
+
+#### Metrics
+
+A new `PolicyReportEvaluation` metric has to be created akin to the current `PolicyEvaluation`
+metric. The reason for not adding `mode` to the existing `PolicyEvaluation` is to keep alerting
+queries for true evaluations easy (without the need to filter `mode = "enforce"`). The metric will
+have the same baggage as the `PolicyEvaluation` metric.
```

Comment:
What about the other queries? I wouldn't want to come up with duplicating each data series we create.

Can't we just filter out by baggage?

Reply:
I would keep things simple at this time. Evaluations is what let the user understand what this policy would be rejecting had it been in `enforce` mode. Our policy count in the controller won't see any change, we can add the `mode` on the baggage, but not a different metric.

I consider evaluations need a different metric because it would impact graphs for things that are not really rejected.

Reply:
My 50 cents on the discussion... If added the `mode` in the baggage we can create a variable that allows users to switch among modes just by changing one field. Something similar to what we did for the graphs showing data for specific policies. So, AFAICS, the user will not need to rewrite the query.

Reply:
Ok, I see we're talking just about one metric. Still, I think we should use the same metric and eventually update the grafana dashboard to filter out the rejections that are "coming" from a policy in "report" mode

Reply:
Honestly, I don't have a blocker here. My only point here is that we don't dirt the evaluations metric in such a way that for getting the metric for final evaluations the user now needs to filter out a boolean, so they can ignore policies on report mode.

Reply:
+1, I'm updating the RFC to remove the new metric. I started my reply before you wrote @flavio.

Reply:
Updated.